### PR TITLE
PRIME 2264 - Next Steps Changes

### PR DESCRIPTION
**Pull request by Paahn**
- a new page/step, Next Step, is added and it is able to send out email to multiple email address and multiple care setting.
- PharmaNet Summary page (Share GPID/Approval from left navigation) is updated to match the Next Step page.  

Note: some test cases to be considered
- New enrolment with single and multiple care settings, then send provisioner email 
- Enrolment renewal/profile update with single and multiple care settings, then send provisioner email 
- Go to "Share GPID/Approval" page and send provisioner email

https://pr-2189.pharmanetenrolment.gov.bc.ca/info

### CODE REVIEW

**Comment by bergomi02** [comment]
@neophyte57 When I updated the code, I noticed the new component for next-step functions very similarly to the existing pharmanet-enrolment-summary component. Please let me know if you think we should combine the components into one. In the current state, the pharmanet-enrolment-summary works for both situations, "Next Step" and "Share GPID". Since it was implemented as such, I tried to change the code as little as possible. We talked about it when you have time to review this PR.

File: prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts [hunk 1]
```diff
@@ -0,0 +1,330 @@
+import { ChangeDetectorRef, Component, OnInit } from '@angular/core';
+import { FormArray, FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
+import { MatDialog } from '@angular/material/dialog';
+import { ActivatedRoute, Router } from '@angular/router';
+import { AuthService } from '@auth/shared/services/auth.service';
+import { ConfigModule } from '@config/config.module';
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts:6]: There are several unused imports

File: prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+    private fb: FormBuilder,
+    protected enrolmentService: EnrolmentService,
+    protected enrolmentResource: EnrolmentResource,
+    protected enrolmentFormStateService: EnrolmentFormStateService,
+    protected toastService: ToastService,
+    protected logger: ConsoleLoggerService,
+    protected utilService: UtilsService,
+    protected formUtilsService: FormUtilsService,
+    protected authService: AuthService,
+    private configService: ConfigService,
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts:72]: `configService` appears to be unused

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.scss [hunk 3]
```diff
@@ -13,3 +13,9 @@
 button mat-icon{
   font-size: 1.2rem;
 }
+
+.vc {
+  display: flex;
+  flex-direction: column;
+  margin-top: 2rem;
+}
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.scss:21]: Is this used?

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts [hunk 4]
```diff
@@ -137,62 +140,25 @@ export class PharmanetEnrolmentSummaryComponent extends BaseEnrolmentPage implem
...
-      }
-      case this.CareSettingEnum.HEALTH_AUTHORITY: {
-        formControl = this.healthAuthorityEmails;
-        break;
+  public sendProvisionerAccessLinkTo() {
+    let valid = true;
+    this.careSettingConfigs.forEach((config) => {
+      valid = valid && config.formArray.valid;
+      if (!config.formArray.valid) {
+        config.formArray.markAllAsTouched();
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:152]: Why is this line needed?

File: prime-angular-frontend/src/app/shared/models/email-pair.model.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+export interface EmailPair {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/models/email-pair.model.ts:1]: Can this be named `EmailsForCareSetting` or something more descriptive?

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts [hunk 6]
```diff
@@ -263,8 +240,43 @@ export class PharmanetEnrolmentSummaryComponent extends BaseEnrolmentPage implem
     }
   }
 
+  public GetEmailsGroup(careSettingCode: number) {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:243]: Please add return type information; that's the project's convention despite compiler acceptance

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts [hunk 7]
```diff
@@ -47,7 +48,8 @@ export class PharmanetEnrolmentSummaryComponent extends BaseEnrolmentPage implem
     setting: string,
     settingPlural: string,
     settingCode: number,
-    formControl: FormControl,
+    formArray: FormArray,
+    formArrayName: string,
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:52]: Where is `formArrayName` used?

File: prime-dotnet-webapi/Models/Api/EmailPair.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+namespace Prime.Models.Api
+{
+    public class EmailPair
```
Comment (neophyte57) [inline prime-dotnet-webapi/Models/Api/EmailPair.cs:3]: More descriptive name please?

File: prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<ng-container [formGroup]="form">
+  <app-form-icon-group (event)="removeEmailInput()"
+                       [show]="showRemoveButton">
+    <div class="row"
+         testing>
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html:5]: Is `testing` needed?

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts (hunk 4 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:155]: Maybe a better way to do it, but it is needed. It checks if all email address form validation is valid. If it's not, it will trigger the UI validation by .markAllAsTouched().

File: prime-angular-frontend/src/app/shared/models/email-pair.model.ts (hunk 5 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/shared/models/email-pair.model.ts:1]: Good idea. Naming is updated.

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts (hunk 6 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:246]: fixed

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts [hunk 10]
```diff
@@ -59,16 +61,17 @@ export class PharmanetEnrolmentSummaryComponent extends BaseEnrolmentPage implem
     private enrolmentResource: EnrolmentResource,
     private enrolmentService: EnrolmentService,
     private dialog: MatDialog,
-    private toastService: ToastService
+    private toastService: ToastService,
+    private configService: ConfigService
```
[the remark this answers is not in the document]
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:68]: removed

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.scss (hunk 3 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.scss:21]: removed

File: prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts (hunk 2 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts:76]: removed

File: prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts (hunk 1 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/next-steps/next-steps.component.ts:6]: removed

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts (hunk 4 above)
Comment (neophyte57) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:153]: IMO, worth adding as a comment

File: prime-angular-frontend/src/app/shared/models/email-pair.model.ts (hunk 5 above)
Comment (neophyte57) [reply prime-angular-frontend/src/app/shared/models/email-pair.model.ts]: Can you also please rename the file accordingly?

File: prime-dotnet-webapi/Controllers/ProvisionerAccessController.cs [hunk 11]
```diff
@@ -84,18 +84,23 @@ public async Task<ActionResult> GetAccessTokens()
...
         [ProducesResponseType(typeof(ApiMessageResponse), StatusCodes.Status400BadRequest)]
         [ProducesResponseType(StatusCodes.Status401Unauthorized)]
         [ProducesResponseType(StatusCodes.Status403Forbidden)]
         [ProducesResponseType(typeof(ApiResultResponse<EnrolmentCertificateAccessToken>), StatusCodes.Status201Created)]
-        public async Task<ActionResult> SendProvisionerLink(int enrolleeId, int careSettingCode, FromBodyText providedEmails)
+        public async Task<ActionResult> SendProvisionerLink(int enrolleeId, [FromBody] EmailPair[] providedEmails)
         {
-            var emails = Email.ParseCommaSeparatedEmails(providedEmails);
-            if (!emails.Any())
+            // var emails = Email.ParseCommaSeparatedEmails(providedEmails);
```
Comment (neophyte57) [inline prime-dotnet-webapi/Controllers/ProvisionerAccessController.cs:96]: Please remove commented code

File: prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+import { ViewportService } from '@core/services/viewport.service';
+import { FormControlValidators } from '@lib/validators/form-control.validators';
+
+import { NextStepsFormState } from '@paper-enrolment/pages/next-steps-page/next-steps-form-state.class';
+@Component({
+  selector: 'app-email-form',
+  templateUrl: './email-form.component.html',
+  styleUrls: ['./email-form.component.scss']
+})
+export class EmailFormComponent implements OnInit {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.ts:14]: Wondering if this would be better named `EmailInputComponent` or `EmailFieldComponent`

File: prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+      <div class="col">
+
+        <mat-form-field class="w-100">
+          <input matInput
+                 placeholder="Email"
+                 formControlName="email"
+                 maxlength="50">
+          <mat-error *ngIf="email.errors?.['required']">Required</mat-error>
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html:12]: The convention used in the project seems to be `email.hasError('required')`

File: prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.ts (hunk 12 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.ts:14]: "EmailFormComponent" is consistent with other form components under the same structure.
image

File: prime-dotnet-webapi/Controllers/ProvisionerAccessController.cs (hunk 11 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Controllers/ProvisionerAccessController.cs:96]: removed

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts [hunk 14]
```diff
@@ -278,47 +292,73 @@ export class PharmanetEnrolmentSummaryComponent extends BaseEnrolmentPage implem
...
+            formArrayName: 'deviceProviderEmails',
+            subheaderContent: `Send your approval to your device provider's PharmaNet administrator (e.g. office manager). If you work in more than one clinic make sure you include every PharmaNet administrator's email. Your PharmaNet administrator(s) will contact you once your PharmaNet access has been set up.`
           };
         }
       }
     });
   }
 
-  private buildVendorEmailGroup(): FormGroup {
+  protected addEmail(emailsGroup: FormArray, email?: string): void {
```
Comment (neophyte57) [inline prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:334]: Rather than `emailsGroup`, is `emailsArray` better?

File: prime-angular-frontend/src/app/shared/models/email-pair.model.ts (hunk 5 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/shared/models/email-pair.model.ts]: ops, file renamed.

File: prime-dotnet-webapi/Models/Api/EmailPair.cs (hunk 8 above)
Comment (bergomi02) [reply prime-dotnet-webapi/Models/Api/EmailPair.cs:3]: right, the class name and the file name are updated.

File: prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html (hunk 13 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/shared/components/forms/email-form/email-form.component.html]: code is updated.

File: prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts (hunk 14 above)
Comment (bergomi02) [reply prime-angular-frontend/src/app/modules/enrolment/pages/pharmanet-enrolment-summary/pharmanet-enrolment-summary.component.ts:334]: hm, kindda picky. I guess why not.

**Review by neophyte57 — APPROVED** [review]
Thanks @bergomi02 for suffering the overly long code review.  🎇
